packages/salesforcedx-vscode-replay-debugger/src/states/variableAssignmentState.ts

```diff
@@ -26,106 +26,141 @@ export class VariableAssignmentState implements DebugLogState {
-          const refContainer = logContext.getRefsMap().get(ref)!;
-          container.variables = refContainer.variables;
-          container.variablesRef = refContainer.variablesRef;
-          container.name = varName;
+      if (ref) {
+        // update the refcontainer mapping
+        if (!refMap.has(ref)) {
+          logContext
+            .getRefsMap()
+            .set(ref, new ApexVariableContainer('', '', '', ref));
         }
-        if (value.indexOf('{') === 0 && value !== '{}') {
-          container.value = '';
-          container.variablesRef = logContext
-            .getVariableHandler()
-            .create(container);
-          this.parseAndPopulate(value, container);
+        const refContainer = refMap.get(ref)!;
+        if (value !== '{}') {
+          if (isNested) {
+            if (value.indexOf('{') === 0) {
+              const topLevel = new ApexVariableContainer(varName, '', '');
+              refContainer.variables.set(varName, topLevel);
+              topLevel.variablesRef = logContext
+                .getVariableHandler()
+                .create(topLevel);
+              this.parseJSONAndPopulate(value, topLevel, logContext);
+            } else {
+              if (refMap.has(value)) {
+                const pulledRef = refMap.get(value) as ApexVariableContainer;
+                const tmpContainer = new ApexVariableContainer(
+                  varName,
+                  pulledRef.value,
+                  pulledRef.type,
+                  value
+                );
+                tmpContainer.variables = pulledRef.variables;
+                tmpContainer.variablesRef = logContext
+                  .getVariableHandler()
+                  .create(tmpContainer);
+                refContainer.variables.set(varName, tmpContainer);
+              } else if (refContainer.variables.has(varName)) {
+                const varContainer = refContainer.variables.get(
+                  varName
+                ) as ApexVariableContainer;
+                varContainer.value = value;
+              } else {
+                refContainer.variables.set(
+                  varName,
+                  new ApexVariableContainer(varName, value, '')
+                );
+              }
+            }
+          } else if (value.indexOf('{') === 0) {
+            this.parseJSONAndPopulate(value, refContainer, logContext);
+          } else {
+            refContainer.variables.set(
+              varName,
+              new ApexVariableContainer(varName, value, '')
+            );
+          }
         }
-        // assigning to variable's fields, currently only going to work for a variable in local
-      } else if (name.indexOf('.') !== -1 && ref !== '0') {
-        container = logContext.getRefsMap().get(ref)!;
-        if (container) {
+
+        // update the correct toplevel container
+        if (
+          container &&
+          !container.type.startsWith('Map<') &&
```

Comment:
This just looks odd to me. We're looking to see whether or not the collection is generic which is going to have the pattern of <SomeString>, can't we just match that? I'd consider moving those 3 checks into an their own call.